### PR TITLE
Tickets/dm 31852

### PR DESCRIPTION
**Pull request by wvreeven**

### CODE REVIEW

File: python/lsst/ts/ess/common/abstract_command_handler.py [hunk 1]
```diff
@@ -78,6 +80,8 @@ def __init__(self, callback: typing.Callable, simulation_mode: int) -> None:
         self._configuration: typing.Optional[typing.Dict[str, typing.Any]] = None
         self._started = False
 
+        self._devices: typing.List[BaseDevice] = []
```
Comment (r-owen) [inline python/lsst/ts/ess/common/abstract_command_handler.py:83]: Regarding the fact that this constructor is abstract (lines 68 and 69) above):

I suggest making all your `__init__` methods concrete, unless you have the special case that the abstract base class has no other abstract methods, yet you still feel it must be abstract. Making `__init__` abstract is an annoyance because subclasses for which the base class constructor suffices must still define their own.

This particular class has no other abstract methods so it potentially is one of these special cases. But you also have a mock subclass of this class that has no additional behavior, suggesting that this base class could usefully be a _concrete_ class.

You have several other abstract classes that have other abstract methods, and so don't need `__init__` to be abstract.

File: python/lsst/ts/ess/common/abstract_command_handler.py [hunk 2]
```diff
@@ -196,10 +206,97 @@ async def stop_sending_telemetry(self) -> None:
...
+            self.log.debug(f"Closing {device} device with name {device.name}")
+            await device.close()
+
         self._started = False
 
-    @abstractmethod
-    async def disconnect_devices(self) -> None:
-        """Loop over the configuration and start all devices."""
-        raise NotImplementedError
+    def get_device(
```
Comment (r-owen) [inline python/lsst/ts/ess/common/abstract_command_handler.py:217]: This creates a new device rather than obtaining a cached device. If that is how you anticipate subclasses working then please consider naming this `make_device`.

I am surprised by this behavior because you have an instance variable `_devices`, but this method seems to not interact with it in any way.

File: python/lsst/ts/ess/common/abstract_command_handler.py [hunk 3]
```diff
@@ -196,10 +206,97 @@ async def stop_sending_telemetry(self) -> None:
...
+            return sensor
+        elif device_configuration[Key.SENSOR_TYPE] == SensorType.WIND:
+            sensor = WindSensor(
+                log=self.log,
+            )
+            return sensor
+        raise RuntimeError(
+            f"Could not get a {device_configuration[Key.SENSOR_TYPE]!r} sensor. "
+            "Please check the configuration."
+        )
```
Comment (r-owen) [inline python/lsst/ts/ess/common/abstract_command_handler.py:302]: I very strongly suggest that you replace this with a sensor registry: a global dict of SensorType: sensor class.

The current code does not scale well to adding new sensor types and it puts sensor-class-specific knowledge inside of this class, which is unnecessary. Furthermore the long if/else is inefficient and error-prone.

Here's the basic pattern. In a file such as `sensor_registry.py` define:
```
SensorRegistry = dict()

def register_sensor(sensor_type, sensor_class):
    """Add a sensor class (subclass of BaseSensor) to the SensorRegistry.

    ...
    """
    sensor_type = SensorType(sensor_type)
    if not isinstance(sensor_class, BaseSensor):
        raise TypeError(...)
    if sensor_type in SensorRegistry:
        raise ValueError(...)  # duplicate
    SensorRegistry[sensor_type] = sensor_class
```
Then each sensor class you define *registers itself*, which is what makes this scale so well to new sensors. For example in `foo_sensor.py`:
```
from sensor_registry import register_sensor

class FooSensor(BaseSensor):
    ...

register_sensor(SensorType.FOO_SENSOR, FooSensor)
```
Then to create a sensor all you have to do is the following (plus possibly a bit of error handling, if the exceptions that will already be raised aren't clear enough):
```
def make_sensor(device_configuration, log):
    sensor_type = device_configuration[Key.SENSOR_TYPE]
    sensor_class = sensor_registry[sensor_type]
    return sensor_class(log=log)
```

File: python/lsst/ts/ess/common/abstract_command_handler.py [hunk 4]
```diff
@@ -196,10 +206,97 @@ async def stop_sending_telemetry(self) -> None:
...
+            name=device_configuration[Key.NAME],
+            device_id=device_configuration[Key.FTDI_ID],
+            sensor=sensor,
+            callback_func=self._callback,
+            log=self.log,
+            disconnected_channel=None,
+        )
+        return device
+
+    def get_sensor(
```
Comment (r-owen) [inline python/lsst/ts/ess/common/abstract_command_handler.py:257]: Please consider:
* Renaming this to `make_sensor` (it creates a sensor, it doesn't get one that has been created) 
* Making it a free function (the only thing it uses from the class is the log). I suggest making a sensor registry below, and this could easily become a free function in the module that defines the registry.

File: python/lsst/ts/ess/common/device/mock_device.py [hunk 5]
```diff
@@ -0,0 +1,188 @@
...
+            return ""
+
+        prefix = f"C{i:02d}="
+        value = random.uniform(MockTemperatureConfig.min, MockTemperatureConfig.max)
+        if i == self._disconnected_channel:
+            value = float(DISCONNECTED_VALUE)
+        return f"{prefix}{value:09.4f}"
+
+    def _format_hbx85_humidity(self, index: int) -> str:
+        if index < self._missed_channels:
```
[the remark this answers is not in the document]
Comment (wvreeven) [reply python/lsst/ts/ess/common/device/mock_device.py:113]: This test is necessary where it is, despite the repetition, because moving it elsewhere would lead to even more repetition.

File: python/lsst/ts/ess/common/device/mock_device.py [hunk 6]
```diff
@@ -0,0 +1,188 @@
...
+            A string representing a temperature.
+        """
+        if i < self._missed_channels:
+            return ""
+
+        prefix = f"C{i:02d}="
+        value = random.uniform(MockTemperatureConfig.min, MockTemperatureConfig.max)
+        if i == self._disconnected_channel:
+            value = float(DISCONNECTED_VALUE)
+        return f"{prefix}{value:09.4f}"
```
[the remark this answers is not in the document]
Comment (wvreeven) [reply python/lsst/ts/ess/common/device/mock_device.py:110]: I agree with your concern. The code was basically moved from ts_ess_controller to this project and was left untouched otherwise. Would you be OK with it if I opened another ticket to implement this and your other concern further down?

File: python/lsst/ts/ess/common/sensor/base_sensor.py [hunk 7]
```diff
@@ -0,0 +1,71 @@
...
+from abc import ABC, abstractmethod
+import logging
+from typing import List
+
+
+class BaseSensor(ABC):
+    """Base class for the different types of Sensors.
+
+    This class holds all common code for the sensors. Sensor specific code
+    (for instance temperature or wind sensor) needs to be implemented in a
```
[the remark this answers is not in the document]
Comment (wvreeven) [reply python/lsst/ts/ess/common/sensor/base_sensor.py:33]: Done

File: python/lsst/ts/ess/common/sensor/base_sensor.py [hunk 8]
```diff
@@ -0,0 +1,71 @@
...
+    ----------
+    log: `logging.Logger`
+        The logger to create a child logger for.
+    num_channels: `int`, optional
+        The number of channels that the sensor will produce telemetry for. The
+        default value is 0 meaning that the number of channels is not variable.
+    """
+
+    @abstractmethod
+    def __init__(self, log: logging.Logger, num_channels: int = 0) -> None:
```
[the remark this answers is not in the document]
Comment (wvreeven) [reply python/lsst/ts/ess/common/sensor/base_sensor.py:46]: Done. I did this because I am used to the way Java handles abstract classes and prevents initialization even if there are no abstract methods in them. I understand that Python handles this differently so I am adapting my coding style to that.

File: python/lsst/ts/ess/common/abstract_command_handler.py (hunk 1 above)
Comment (wvreeven) [reply python/lsst/ts/ess/common/abstract_command_handler.py:83]: Done. Again, I did this because I am used to the way Java handles abstract classes and prevents initialization even if there are no abstract methods in them. I understand that Python handles this differently so I am adapting my coding style to that.

File: python/lsst/ts/ess/common/device/base_device.py [hunk 9]
```diff
@@ -0,0 +1,167 @@
...
+    async def _run(self) -> None:
+        """Run sensor read loop.
+
+        If enabled, loop and read the sensor and pass result to callback_func.
+        """
+        self.log.debug("Starting sensor.")
+        while not self._telemetry_loop.done():
+            self.log.debug("Reading data.")
+            curr_tai: float = utils.current_tai()
+            response: int = ResponseCode.OK
```
[the remark this answers is not in the document]
Comment (wvreeven) [reply python/lsst/ts/ess/common/device/base_device.py:110]: I added the type annotations because that way my IDE was able to do code completion and without not. Is it a big issue if I left the type annotations there?

File: python/lsst/ts/ess/common/device/base_device.py [hunk 10]
```diff
@@ -0,0 +1,167 @@
...
+        await self.basic_open()
+        self.is_open = True
+
+        self.log.debug(f"Starting read loop for {self.name!r} sensor.")
+        self._telemetry_loop = asyncio.create_task(self._run())
+
+    @abstractmethod
+    async def basic_open(self) -> None:
+        """Open the Sensor Device."""
+        pass
```
[the remark this answers is not in the document]
Comment (wvreeven) [reply python/lsst/ts/ess/common/device/base_device.py:99]: Done

File: tests/test_mock_hb85a_device.py [hunk 11]
```diff
@@ -0,0 +1,105 @@
...
+            name=md_props.name,
+            device_id="MockDevice",
+            sensor=sensor,
+            callback_func=self._callback,
+            log=self.log,
+            missed_channels=md_props.missed_channels,
+            in_error_state=md_props.in_error_state,
+        )
+
+        await mock_device.open()
```
Comment (r-owen) [inline tests/test_mock_hb85a_device.py:60]: I strongly suggest (on this ticket or a later ticket) that you make devices async context managers. This would allow you to write your tests as:
```
async with common.device.MockDevice(...) as mock_device:
    # Use your mock device here. Do NOT open or close it.
```
Users don't have to remember to open or close a device, and if a test fails, the device will still be closed (unlike your current tests, though I realize that probably doesn't really matter for mock devices).

It's a trivial change. Just add these two methods to `BaseDevice`:
```
async def __aenter__(self) -> BaseDevice:
    await self.open()
    return self

async def __aexit__(
    self,
    type: typing.Optional[typing.Type[BaseException]],
    value: typing.Optional[BaseException],
    traceback: typing.Optional[types.TracebackType],
) -> None:
    await self.close()
```

File: tests/test_socket_server.py [hunk 12]
```diff
@@ -32,24 +32,20 @@
     format="%(asctime)s:%(levelname)s:%(name)s:%(message)s", level=logging.DEBUG
 )
 
-# Standard timeout in seconds.
 TIMEOUT = 5
+"""Standard timeout in seconds."""
```
Comment (r-owen) [inline tests/test_socket_server.py:36]: I suggest you revert this change.

Or have you found a way to make sphinx actually use the doc strings after module constants? If you have, please tell me how to do it.

File: tests/test_socket_server.py [hunk 13]
```diff
@@ -173,13 +136,29 @@ async def check_server_test(self, num_channels: int) -> None:
...
         self.assertEqual(common.ResponseCode.OK, data[common.Key.RESPONSE])
 
-        reply = await self.read()
-        reply_to_check = reply[common.Key.TELEMETRY]
-        self.check_temperature_reply(reply=reply_to_check, num_channels=num_channels)
+        # Make sure that the mock sensor outputs data for a disconnected
+        # channel.
+        self.srv.command_handler._devices[
+            0
+        ]._disconnected_channel = md_props.disconnected_channel
```
Comment (r-owen) [inline tests/test_socket_server.py:143]: Please don't use private variables outside of the class. If you really want access to the devices then the instance variable should be named `devices` and the disconnected channel should be named `disconnected_channel` (or the device should have a public way of setting the private variable, such as a method or property setter). Similarly for the other mock device extra variables (_missed_channels, etc.).

Renaming these to eliminate the leading _ would be a pretty quick global search/replace. My only concern is if any of the mock-device-specific variables could ever collide with the base class. You could eliminate the risk by having a dataclass or other struct to hold the mock-specific configuration (which would be more work). But the ones I looked at don't seem likely to collide anyway.

File: python/lsst/ts/ess/common/device/base_device.py [hunk 14]
```diff
@@ -0,0 +1,167 @@
...
+                line = f"{self._sensor.terminator}"
+                response = ResponseCode.DEVICE_READ_ERROR
+
+            if self._in_error_state:
+                response = ResponseCode.DEVICE_READ_ERROR
+
+            sensor_telemetry: List[float] = await self._sensor.extract_telemetry(
+                line=line
+            )
+            output: List[Union[str, float, int]] = [
```
[the remark this answers is not in the document]
Comment (wvreeven) [reply python/lsst/ts/ess/common/device/base_device.py:123]: As it turns out, my code actually only has one copy of this particular type description. All others are unions between `str` and `int` or `str` and `float` but not all three. I simply removed the type annotation because it doesn't seem to be necessary at all.

File: tests/test_socket_server.py (hunk 13 above)
Comment (wvreeven) [reply tests/test_socket_server.py:143]: I renamed these, and several other, variables and removed the leading `_` from the names.

File: python/lsst/ts/ess/common/sensor/omega_hx85ba.py [hunk 15]
```diff
@@ -0,0 +1,109 @@
...
+
+        Parameters
+        ----------
+        line: `str`
+            A line of comma separated telemetry.
+
+        Returns
+        -------
+        output: `list`
+            A list containing the telemetry as measured by the sensor.
```
[the remark this answers is not in the document]
Comment (wvreeven) [reply python/lsst/ts/ess/common/sensor/omega_hx85ba.py:86]: Done

File: python/lsst/ts/ess/common/test_utils.py [hunk 16]
```diff
@@ -0,0 +1,160 @@
...
+        num_channels: int = 0,
+        missed_channels: int = 0,
+        disconnected_channel: int = None,
+        in_error_state: bool = False,
+    ) -> None:
+        self.name: str = name
+        self.num_channels = num_channels
+        self.missed_channels = missed_channels
+        self.disconnected_channel = disconnected_channel
+        self.in_error_state = in_error_state
```
[the remark this answers is not in the document]
Comment (wvreeven) [reply python/lsst/ts/ess/common/test_utils.py:61]: I'll do that as part of DM-32284.

File: tests/test_mock_hb85a_device.py (hunk 11 above)
Comment (wvreeven) [reply tests/test_mock_hb85a_device.py:60]: This is a really cool suggestion, thanks! I ran into an issue where the type annotation for `__aenter__` couldn't resolve `BaseDevice` which can be solved in two ways: either make the type annotation a string `"-> BaseDevice"` or add `from __future__ import annotations`. I chose the latter but wonder if the former would be acceptable as well?

File: python/lsst/ts/ess/common/device/mock_device.py (hunk 6 above)
[the remark this answers is not in the document]
Comment (wvreeven) [reply python/lsst/ts/ess/common/device/mock_device.py]: I decided to do this in this ticket anyway.

File: python/lsst/ts/ess/common/abstract_command_handler.py (hunk 4 above)
Comment (wvreeven) [reply python/lsst/ts/ess/common/abstract_command_handler.py]: I decided to do this in this ticket anyway.

File: python/lsst/ts/ess/common/device/mock_device.py (hunk 6 above)
[the remark this answers is not in the document]
Comment (wvreeven) [reply python/lsst/ts/ess/common/device/mock_device.py]: I decided to do this in this ticket anyway.

File: python/lsst/ts/ess/common/device/mock_device.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (wvreeven) [reply python/lsst/ts/ess/common/device/mock_device.py]: I decided to keep the tests in the individual methods as I don't see a simple way of centralizing this. But I am open for suggestions.

File: python/lsst/ts/ess/common/test_utils.py (hunk 16 above)
[the remark this answers is not in the document]
Comment (wvreeven) [reply python/lsst/ts/ess/common/test_utils.py]: I decided to leave this code as is.
